shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizActionDispatcher.kt

```diff
@@ -146,6 +151,29 @@ class StepQuizActionDispatcher(
                     onNewMessage(Message.CreateAttemptSuccess(action.step, action.attempt, submissionState))
                 }
             }
+            is Action.CreateSubmissionCheckLimit -> {
+                val isFreemiumEnabled = freemiumInteractor
+                    .isFreemiumEnabled()
+                    .getOrElse {
+                        return onNewMessage(Message.CreateSubmissionCheckLimitResult.NetworkError)
+                    }
+
+                if (isFreemiumEnabled) {
+                    val subscription = currentSubscriptionStateRepository
+                        .getState()
+                        .getOrElse {
+                            return onNewMessage(Message.CreateSubmissionCheckLimitResult.NetworkError)
+                        }
+
+                    if (subscription.isFree && subscription.stepsLimitLeft == 0) {
+                        return onNewMessage(Message.ShowProblemsLimitReachedModal)
+                    }
+                }
```

Comment:
In my opinion, it's better to move this logic into `FreemiumInteractor.isSubmissionAvailable()` method.